src/main.cc

```diff
@@ -1437,9 +1437,52 @@ StartUsingConfig()
     }
 }
 
+/// register all known modules for handling future RegisteredRunner events
+static void
+RegisterModules()
+{
+    // These registration calls do not represent a RegisteredRunner "event". The
+    // modules registered here should be initialized later, during those events.
+
+    // RegisteredRunner event handlers should not depend on handler call order
+    // and, hence, should not depend on the registration call order below.
+
+    CallRunnerRegistrator(ClientDbRr);
+    CallRunnerRegistrator(CollapsedForwardingRr);
+    CallRunnerRegistrator(IpcIoRr);
+    CallRunnerRegistrator(MemStoreRr);
+    CallRunnerRegistrator(PeerPoolMgrsRr);
+    CallRunnerRegistrator(SharedMemPagesRr);
+    CallRunnerRegistrator(SharedSessionCacheRr);
+    CallRunnerRegistrator(TransientsRr);
+    CallRunnerRegistratorIn(Dns, ConfigRr);
+    CallRunnerRegistratorIn(Rock, SwapDirRr);
+
+#if USE_SQUID_ESI && HAVE_LIBEXPAT
+    CallRunnerRegistratorIn(Esi, ExpatRr);
+#endif
```

Comment:
Guards like these are the biggest drawback of this solution IMO. RegisterModules() has to know what are the preconditions for the corresponding module registrator existence. The only solution for this problem (besides the buggy static initialization-based one that this PR removes), is automated generation of module registration code at Squid build time, when the list of enabled (to build) modules becomes known. That solution requires a lot of work and will complicate this code comprehension further. I do not think it is worth it at this point, but we can consider it if these module registration guards go out of hand.